### PR TITLE
Problem: keep recent snapshots are not optional

### PR DESCRIPTION
**Pull request by mmsqe**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/db.go [hunk 1]
```diff
@@ -116,33 +121,54 @@ func (db *DB) Commit(changeSets []*NamedChangeSet) ([]byte, int64, error) {
...
 		}
 	}
+	return false, nil
+}
+
+// Commit wraps `MultiTree.ApplyChangeSet` to add some db level operations:
+// - manage background snapshot rewriting
+// - write WAL
+func (db *DB) Commit(changeSets []*NamedChangeSet) ([]byte, int64, error) {
+	db.cleanupSnapshotRewrite()
```
Comment (yihuang) [inline memiavl/db.go:171]: This error return need to be handled I guess.

File: memiavl/db.go [hunk 2]
```diff
@@ -116,33 +121,56 @@ func (db *DB) Commit(changeSets []*NamedChangeSet) ([]byte, int64, error) {
...
-				return nil, 0, fmt.Errorf("catchup failed: %w", err)
+				return true, fmt.Errorf("catchup failed: %w", err)
 			}
 			if err := db.reloadMultiTree(result.mtree); err != nil {
-				return nil, 0, fmt.Errorf("switch multitree failed: %w", err)
+				return true, fmt.Errorf("switch multitree failed: %w", err)
 			}
 			// prune the old snapshots
+			// wait until last prune finish
+			db.pruneSnapshotLock.Lock()
```
[the remark this answers is not in the document]
Comment (mmsqe) [reply memiavl/db.go:139]: Commit will block, otherwise we have concurrent issue

[the remark this answers is not in the document]
Comment (yihuang) [reply memiavl/db.go:139]: we can simply put the lock into the go-routine itself?

[the remark this answers is not in the document]
Comment (mmsqe) [reply memiavl/db.go:139]: yes but still not good to pend unfinished things

[the remark this answers is not in the document]
Comment (yihuang) [reply memiavl/db.go:139]: 👌

File: memiavl/db.go [hunk 3]
```diff
@@ -104,10 +112,10 @@ func (db *DB) ApplyUpgrades(upgrades []*TreeNameUpgrade) error {
...
 
-// Commit wraps `MultiTree.ApplyChangeSet` to add some db level operations:
-// - manage background snapshot rewriting
-// - write WAL
-func (db *DB) Commit(changeSets []*NamedChangeSet) ([]byte, int64, error) {
+// cleanupSnapshotRewrite cleans up the old snapshots and switches to a new multitree
+// if a snapshot rewrite is in progress. It returns true if a snapshot rewrite has completed
+// and false otherwise, along with any error encountered during the cleanup process.
+func (db *DB) cleanupSnapshotRewrite() (bool, error) {
 	if db.snapshotRewriteChan != nil {
```
Comment (yihuang) [inline memiavl/db.go:119]: It looks better to quit early when `if db.snapshotRewriteChan == nil {`, instead of a big block of if clause.

**Pull request merged**